drivers/console/aarch64/multi_console.S

```diff
@@ -0,0 +1,270 @@
+	 * Function to insert a new console structure into
+	 * the console list. Should usually be called by
+	 * console_<driver>_register implementations. The
+	 * data structure passed will be taken over by the
+	 * console framework and *MUST* be allocated in
+	 * persistent memory (e.g. the data section) -- to
+	 * prevent mistakes of that kind from going undetected,
+	 * we explicitly test that it is not on the stack.
+	 * In : x0 - address of console_t structure
+	 * Out: x0 - 1 on success, 0 on error
+	 * Clobber list: x0, x1, x14
+	 * -----------------------------------------------
+	 */
+func console_register
+	adrp	x1, __STACKS_START__
+	add	x1, x1, :lo12:__STACKS_START__
+	cmp	x0, x1
+	b.lo	not_on_stack
+	adrp	x1, __STACKS_END__
+	add	x1, x1, :lo12:__STACKS_END__
+	cmp	x0, x1
+	b.hs	not_on_stack
+	mov	x0, #0
+	ret
+not_on_stack:
+	adrp	x14, console_list
+	ldr	x1, [x14, :lo12:console_list]	/* X1 = first struct in list */
+	str	x0, [x14, :lo12:console_list]	/* list head = new console */
+	str	x1, [x0, #CONSOLE_T_NEXT]	/* new console next ptr = X1 */
+	mov	x0, #1
+	ret
+endfunc console_register
+
+	/* -----------------------------------------------
+	 * int console_unregister(console_t *console)
+	 * Function to find a specific console in the list
+	 * of currently active consoles and remove it.
+	 * In: x0 - address of console_t struct to remove
+	 * Out: x0 - removed address, or NULL if not found
+	 * Clobber list: x0, x1, x14
+	 * -----------------------------------------------
+	 */
+func console_unregister
+	adrp	x14, console_list
+	add	x14, x14, :lo12:console_list	/* X14 = ptr to first struct */
+	ldr	x1, [x14]			/* X1 = first struct */
+
+unregister_loop:
+	cbz	x1, unregister_not_found
+	cmp	x0, x1
+	b.eq	unregister_found
+	ldr	x14, [x14]			/* X14 = next ptr of struct */
+	ldr	x1, [x14]			/* X1 = next struct */
+	b	unregister_loop
+
+unregister_found:
+	ldr	x1, [x1]			/* X1 = next struct */
+	str	x1, [x14]			/* prev->next = cur->next */
+	ret
+
+unregister_not_found:
+	mov	x0, #0				/* return NULL if not found */
+	ret
+endfunc console_unregister
+
+	/* -----------------------------------------------
+	 * void console_uninit(void)
+	 * Function to remove all registered consoles.
+	 * Clobber list: x0, x1
+	 * -----------------------------------------------
+	 */
+func console_uninit
+	mov	x1, #0
+	adrp	x0, console_list
+	str	x1, [x0, :lo12:console_list]
+	ret
+endfunc console_uninit
+
+	/* -----------------------------------------------
+	 * int console_switch_state(void)
+	 * Function to switch the current console state.
+	 * The console state determines which of the
+	 * registered consoles are actually used at a time.
+	 * Returns 1 in x0 for convenient use in
+	 * plat_crash_console_init.
+	 * In : x0
+	 * Out: x0 - always 1
+	 * Clobber list: x0, x1
+	 * -----------------------------------------------
+	 */
+func console_switch_state
+	adrp	x1, console_state
+	strb	w0, [x1, :lo12:console_state]
```

Comment:
When implementing this I noticed that it's not really necessary. The value is already implicitly truncated to 8 bits by the strb which is what I picked for CONSOLE_FLAG_SCOPE_MASK (there are only 3 scopes right now, but I wanted to leave room for expansion and clearly mark which bits are for scopes and which could be used for other generic console classification if needed in the future).

I could count the bits but I think it might be useful to leave open the option of doing console_switch_state(0) (effectively turn off all console output for a while) or enable multiple scopes at the same time.

Reply:
You could assert that bits 7:3 are 0 for now and remove/change the assert if new scopes are added later, but I don't feel that strongly about  it. I agree that use of strb at least ensures "non-scope" bits are dropped. I also agree it might be useful to allow turning off all console output.